### DISCUSSION TITLE
Running make -j8 is not producing cardinal-opt file, ModuleNotFoundError: No module named 'TestHarness', and ModuleNotFoundError: No module named 'hit'

### DISCUSSION THREAD

**Discussion by lliw-will** [neams-th-coe/cardinal#893] [Q&A]
I followed the Building Cardinal without MOOSE's conda Environment directions as I require NekRS; however, after running make there is no cardinal-opt despite the process seemingly running in its entirety. There was an error saying GLIBCXX_3.4.32 was not found and there was no module named 'hit'. When I attempted to run the run_tests file I got an error saying no module named 'TestHarness'.

Any help with these problems would be greatly appreciated.

**Comment by aprilnovak** [comment]
Hi @lliw-will, if no `cardinal-opt` was produced it means that the build command was not successful. Oftentimes, the actual error message gets buried in the screen output and won't be immediately obvious. Would you please do the following:

```
cd cardinal
rm -rf build/ install/
make -j8  >> out.txt 2>&1
```

and then attach `out.txt` here?

**Reply by lliw-will** [reply]
Here is the result:
[out.txt](https://github.com/neams-th-coe/cardinal/files/15501780/out.txt)

**Reply by aprilnovak** [reply]
Thanks @lliw-will  - it looks like the issue is related to `version `GLIBCXX_3.4.32' not found`. Can you try this:

```
cd cardinal/contrib/moose/test
make -j8 >> moose.txt 2>&1
```

and then attach `moose.txt` here? This will build MOOSE independently of Cardinal -- I think the issue is not Cardinal-specific, so this would confirm that (in which case I'll move this conversation over to the MOOSE discussions page).

**Reply by lliw-will** [reply]
Thanks for the help, here is the moose.txt file:
[moose.txt](https://github.com/neams-th-coe/cardinal/files/15502069/moose.txt)

**Reply by aprilnovak** [reply]
Thanks @lliw-will , that looks like you were able to build MOOSE successfully. 

Can you try some of the suggestions on this page? I think you may just need to update your compiler.

https://askubuntu.com/questions/575505/glibcxx-3-4-20-not-found-how-to-fix-this-error

**Reply by lliw-will** [reply]
Hi @aprilnovak I was able to update my compiler and it appears to have gotten rid of the 'GLIBCXX_3.4.32 not found' error; however, the build command is not successful. Here is the results of running make: 
[out.txt](https://github.com/user-attachments/files/15537741/out.txt)

**Reply by aprilnovak** [reply]
Hi @lliw-will, if you search for "error" in the output, you'll see this line early on:

```
CMake Error at CMakeLists.txt:228 (message):
  ENABLE_AMGX required OCCA_CUDA_ENABLED!
```

You will need to edit the Makefile to set `OCCA_CUDA_ENABLED=1`

I'll see if we can update Cardinal's makefile to terminate the build when hitting one of these errors so that they're easier to see.

**Reply by lliw-will** [reply]
That is strange as I have changed the Makefile and set `OCCA_CUDA_ENABLED=1`

**Reply by aprilnovak** [reply]
If you change the build settings, you will need to erase the `build/` and `install/` directories -- every time you re-run `make`, it will re-use any cached files in those folders instead of re-generating them (nominally, this is a good thing because it will save you a lot of time with all further compilations if you are developing code, but it is something that needs to be treated carefully if you change anything in the Makefile).

```
cd cardinal
rm -rf build/ install
make
```

**Reply by lliw-will** [reply]
I had run the `rm -rf build/ install` command before every re-run of `make`

**Reply by aprilnovak** [reply]
Can you please show me your `Makefile` that you are using? In the `out.txt` you attached, it looks like the `OCCA_CUDA_ENABLED` isn't being picked up:

```
-- OCCA_CUDA_ENABLED: 0
-- OCCA_HIP_ENABLED: 
-- OCCA_DPCPP_ENABLED: 0
-- OCCA_OPENCL_ENABLED: 
CMake Error at CMakeLists.txt:228 (message):
  ENABLE_AMGX required OCCA_CUDA_ENABLED!
```

that first line should be 1 instead of 0.

**Reply by lliw-will** [reply]
Here you go I have copied and pasted the code into a txt file as the makefile was not supported: [Makefile.txt](https://github.com/user-attachments/files/15538451/Makefile.txt)

**Reply by aprilnovak** [reply]
Hmm, that looks okay to me. I am really not sure -- can we try one more drastic attempt? Can you try entirely deleting the cardinal directory, re-cloning, and repeating all steps to build? (making sure you edit the `Makefile` as you had already).

**Reply by lliw-will** [reply]
I went a step further and reinstalled kubuntu, I repeated the build instructions including editing the Makefile with no luck. I tried to run `make` with a CPU backend to see if CUDA was an issue. Here are the txt files for the attempted CUDA and CPU builds: 
[out_cpu.txt](https://github.com/user-attachments/files/15571030/out_cpu.txt)
[out_cuda.txt](https://github.com/user-attachments/files/15571033/out_cuda.txt)

**Reply by aprilnovak** [reply]
Thanks @lliw-will this is very helpful. I'll start with the CPU one, which shows something puzzling related to the stochastic tools module not building...

```
Linking Test Library /home/will/cardinal/test/lib/libcardinal_test-opt.la...
Linking Library /home/will/cardinal/lib/libcardinal-opt.la...
Linking Executable /home/will/cardinal/cardinal-opt...
/usr/bin/ld: /home/will/cardinal/contrib/moose/modules/stochastic_tools/lib/libstochastic_tools-opt.so: undefined reference to `SamplerInterface::validParams()'
/usr/bin/ld: /home/will/cardinal/contrib/moose/framework/libmoose-opt.so: undefined reference to `SamplerInterface::SamplerInterface(MooseObject const*)'
collect2: error: ld returned 1 exit status
make: *** [/home/will/cardinal/contrib/moose/framework/app.mk:457: /home/will/cardinal/cardinal-opt] Error 1
```

Just out of curiosity, can you try setting this line in the Makefile:

```
STOCHASTIC_TOOLS    := no
```

and then do

```
rm -rf build/ install/
make
```

**Reply by aprilnovak** [reply]
Also @lliw-will, would you please let me know the outcome of:

```
ls cardinal/contrib/moose/framework/build/unity_src/samplers_Unity.*
```

tag @roystgnr

**Reply by lliw-will** [reply]
Here is the outcome of running make with `STOCHASTIC_TOOLS    := no` : 
[out.txt](https://github.com/user-attachments/files/15655617/out.txt)

`ls cardinal/contrib/moose/framework/build/unity_src/samplers_Unity.*`  

returned: 

cardinal/contrib/moose/framework/build/unity_src/samplers_Unity.C

cardinal/contrib/moose/framework/build/unity_src/samplers_Unity.x86_64-pc-linux-gnu.opt.lo

cardinal/contrib/moose/framework/build/unity_src/samplers_Unity.x86_64-pc-linux-gnu.opt.lo.d

**Reply by roystgnr** [reply]
So you have built the samplers_Unity libtool object files, but the final library is still missing the symbols they should be defining?

Do you get any output from `nm --demangle cardinal/contrib/moose/framework/build/unity_src/.libs/samplers_Unity.* | grep SamplerInterface` ?  Or from `nm --demangle cardinal/contrib/moose/framework/.libs/libmoose*.so` ?

**Reply by lliw-will** [reply]
`nm --demangle cardinal/contrib/moose/framework/build/unity_src/.libs/samplers_Unity.* | grep SamplerInterface`  returns: 
```
0000000000001aa0 T Sampler& SamplerInterface::getSampler<Sampler>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
000000000000016c t Sampler& SamplerInterface::getSampler<Sampler>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) [clone .cold]
0000000000000840 T SamplerInterface::validParams()
0000000000000880 T Sampler& SamplerInterface::getSamplerByName<Sampler>(SamplerName const&)
00000000000017c0 T SamplerInterface::SamplerInterface(MooseObject const*)
00000000000017c0 T SamplerInterface::SamplerInterface(MooseObject const*)
00000000000000fc t SamplerInterface::SamplerInterface(MooseObject const*) [clone .cold]
0000000000000000 V typeinfo for SamplerInterface
0000000000000000 V typeinfo name for SamplerInterface
```

And `nm --demangle cardinal/contrib/moose/framework/.libs/libmoose*.so` returns more text than can be viewed at a time in the terminal, is there a way to write the output to a file if needed?

**Reply by roystgnr** [reply]
Oh, I'm sorry!  I meant to append `| grep SamplerInterface` to the second command too.

**Reply by lliw-will** [reply]
In that case `nm --demangle cardinal/contrib/moose/framework/.libs/libmoose*.so| grep SamplerInterface` returns:
```
0000000001a739d0 T Sampler& SamplerInterface::getSampler<Sampler>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
0000000000e2244c t Sampler& SamplerInterface::getSampler<Sampler>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) [clone .cold]
0000000001a72770 T SamplerInterface::validParams()
0000000001a727b0 T Sampler& SamplerInterface::getSamplerByName<Sampler>(SamplerName const&)
0000000001a736f0 T SamplerInterface::SamplerInterface(MooseObject const*)
0000000001a736f0 T SamplerInterface::SamplerInterface(MooseObject const*)
0000000000e223dc t SamplerInterface::SamplerInterface(MooseObject const*) [clone .cold]
0000000002e8b578 V typeinfo for SamplerInterface
0000000002964f30 V typeinfo name for SamplerInterface
```

**Reply by roystgnr** [reply]
I don't understand this at all.  ld should have no problems with `undefined reference to 'SamplerInterface::SamplerInterface(MooseObject const*)'` if the reference is *right there* - that `T` is for symbols defined in the text section of an object or library file.

Are these messages from two different states?  Maybe you've run other attempts since `out_cpu.txt` and we're seeing the status of those object and library files *now* rather than the status they were in when `make` failed there?  If you try building+linking again, do you still get the linker error?

**Reply by lliw-will** [reply]
@roystgnr are you suggesting `rm -rf build/ install/` and building again or deleting the cardinal repository and cloning it again?
 If you are suggesting `rm -rf build/ install/` this is the result: 
[out2.txt](https://github.com/user-attachments/files/15738963/out2.txt)

**Reply by roystgnr** [reply]
You're now getting a completely different linker error, with `Times` rather than `SamplerInterface`?

I'm not sure what changed, but this is the sort of "two different states" thing I was fearing.

There can't be a race condition in the cardinal build scripts, right?  It's as if something is trying to link to the MOOSE library before the library is done building.

**Reply by aprilnovak** [reply]
I don't think there is a race condition (at least, I'm not aware of any other user encountering behavior like this). For building MOOSE, I think we use the same Makefile contents as all other MOOSE apps (we just have extra instructions to build OpenMC and NekRS).

**Reply by lliw-will** [reply]
I am running Linux (Kubuntu 24.04) on an external SSD and using Microsoft visual studio code, and since the only purpose for this SSD is to run this code I am more than willing to start from absolutely ground zero on a fresh install of whatever Linux distribution you see fit to see if Kubuntu and/or VS code could be part of the problem.

**Reply by aprilnovak** [reply]
Hi @lliw-will, sorry for the delayed response, I was out of office all of last week.

If you're willing to attempt a fresh Linux install, that could be an option to get to the bottom of this. I've always used macOS so I don't have any first-hand advice on which distribution to choose, but I know some folks have used Linux Mint successfully.

**Reply by lliw-will** [reply]
I apologize for the time it took me to respond earlier as I was away for a couple of workshops. 

After installing Mint-cinnamon and going through the build guide for a CPU build the results were as follows: [out_mint_CPU.txt](https://github.com/user-attachments/files/16144472/out_mint_CPU.txt)

**Reply by aprilnovak** [reply]
I recently encountered linking issues from the solid mechanics module, and was able to fix it by deleting Cardinal and starting over with a fresh clone (not a satisfying explanation, but I think that will work for you). Also please make sure you are using the MOOSE hash tied to Cardinal (i.e., you're not setting anything custom for MOOSE_DIR).

**Reply by lliw-will** [reply]
Good news @aprilnovak, I deleted Cardinal and tried building again and the cardinal-opt file was created. This time I had used the main terminal instead of the one used by visual studio code.

Is it possible to build with the CUDA backend without removing the current build?

**Reply by aprilnovak** [reply]
No, MOOSE's build setup does not allow out-of-source builds (so any time a build configuration is changed, the entire `build` directory must be re-created). Fortunately, the most time-consuming process is building PETSc and libMesh, which do NOT need to be repeated if you'd like to add CUDA.

```
cd cardinal
rm -rf build/ install/
# edit Makefile to set OCCA_CUDA_ENABLED=1
make -j8
```
